scripts/grid.js

```diff
@@ -0,0 +1,386 @@
+                color = color_x;
+            } else {
+                color = vec3(0.9);
+            }
+            gl_FragColor = vec4(color, levelAlpha);
+            gl_FragDepth = writeDepth(levelAlpha) ? calcDepth(pos) : 1.0;
+            return;
+        }
+
+        // 1m grid
+        levelPos = pos;
+        levelSize = 1.0 / 100.0;
+        levelAlpha = pristineGrid(levelPos.xz, ddx, ddy, vec2(levelSize)) * fade;
+        if (levelAlpha > epsilon) {
+            gl_FragColor = vec4(vec3(0.7), levelAlpha);
+            gl_FragDepth = writeDepth(levelAlpha) ? calcDepth(pos) : 1.0;
+            return;
+        }
+
+        // 0.1m grid
+        levelPos = pos * 10.0;
+        levelSize = 1.0 / 100.0;
+        levelAlpha = pristineGrid(levelPos.xz, ddx * 10.0, ddy * 10.0, vec2(levelSize)) * fade;
+        if (levelAlpha > epsilon) {
+            gl_FragColor = vec4(vec3(0.7), levelAlpha);
+            gl_FragDepth = writeDepth(levelAlpha) ? calcDepth(pos) : 1.0;
+            return;
+        }
+
+        discard;
+    }
+`;
+
+class Grid extends Script {
+    /**
+     * @type {GraphicsDevice}
+     */
+    _device;
+
+    /**
+     * @type {CameraComponent}
+     * @private
+     */
+    _camera;
+
+    /**
+     * @type {QuadRender}
+     * @private
+     */
+    _quadRender;
+
+    /**
+     * @type {BlendState}
+     */
+    _blendState = new BlendState(
+        true,
+        BLENDEQUATION_ADD, BLENDMODE_SRC_ALPHA, BLENDMODE_ONE_MINUS_SRC_ALPHA,
+        BLENDEQUATION_ADD, BLENDMODE_ONE, BLENDMODE_ONE_MINUS_SRC_ALPHA
+    );
+
+    /**
+     * @type {() => void}
+     */
+    _prerender;
+
+    /**
+     * @type {Vec2}
+     */
+    _halfExtents = new Vec2(Infinity, Infinity);
+
+    /**
+     * @type {Color}
+     * @private
+     */
+    _colorX = new Color(1, 0.3, 0.3);
+
+    /**
+     * @type {Color}
+     * @private
+     */
+    _colorZ = new Color(0.3, 0.3, 1);
+
+    /**
+     * @type {string}
+     */
+    layerName = 'World';
+
+    /**
+     * @param {object} args - The script arguments.
+     */
+    constructor(args) {
+        super(args);
+
+        this._device = this.app.graphicsDevice;
+
+        // create shader
+        const shader = createShaderFromCode(this._device, vertexShader, fragmentShader, 'grid', {
+            vertex_position: SEMANTIC_POSITION
+        });
+        this._quadRender = new QuadRender(shader);
+
+        // set initial size
+        this._set('half_extents', this._halfExtents);
+
+        // set initial colors
+        this._set('color_x', this._colorX);
+        this._set('color_z', this._colorZ);
+    }
+
+    /**
+     * Set the value of a uniform in the shader.
+     *
+     * @param {string} name - The name of the uniform.
+     * @param {Color|Vec3|number} value - The value to set.
+     * @private
+     */
+    _set(name, value) {
+        if (value instanceof Color) {
+            this._device.scope.resolve(name).setValue([value.r, value.g, value.b]);
```

Comment:
@mvaligursky there's no performance impact in polling `resolve()`  here right?

Reply:
resolve is just a lookup into a Map, so not much cost.